### PR TITLE
implement #12231 new flag --xfail-tb

### PR DESCRIPTION
**Pull request by okken**
This implements issue #12231, making xfail tracebacks optional. 
Add `--xfail-tb` flag, which turns traceback output for XFAIL results.

* If the `--xfail-tb` flag is not sent, tracebacks for XFAIL results are NOT shown.
* The style of traceback for XFAIL is set with `--tb`, and can be `auto|long|short|line|native|no`.
* Note: Even if you have `--xfail-tb` set, you won't see them if `--tb=no`.

Some history:

* This is a behavior break, but brings default behavior back to pre-8.0.0 behavior.
* With this change, default `-rx`/ `-ra` behavior is identical to pre-8.0 with respect to xfail tracebacks.
* With pytest 8.0, `-rx` or `-ra` would not only turn on summary reports for xfail, but also report the tracebacks for xfail results.
* This caused issues with some projects that utilize xfail, but don't want to see all of the xfail tracebacks.
* This change detaches xfail tracebacks from `-rx`, and now we turn on xfail tracebacks with `--xfail-tb`.

### CODE REVIEW

File: changelog/12231.feature.rst [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Add `--xfail-tb` flag, which turns traceback output for XFAIL results.
```
Comment (nicoddemus) [inline changelog/12231.feature.rst:1]:
```suggestion
Added `--xfail-tb` flag, which turns on traceback output for XFAIL results.
```

File: changelog/12231.feature.rst [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+* The style of traceback for XFAIL is set with `--tb`, and can be `auto|long|short|line|native|no`.
+* Note: Even if you have `--xfail-tb` set, you won't see them if `--tb=no`.
+
+Some history:
+
+* This is a behavior break, but brings default behavior back to pre-8.0.0 behavior.
+* With this change, default `-rx`/ `-ra` behavior is identical to pre-8.0 with respect to xfail tracebacks.
+* With pytest 8.0, `-rx` or `-ra` would not only turn on summary reports for xfail, but also report the tracebacks for xfail results.
+* This caused issues with some projects that utilize xfail, but don't want to see all of the xfail tracebacks.
+* This change detaches xfail tracebacks from `-rx`, and now we turn on xfail tracebacks with `--xfail-tb`.
```
Comment (nicoddemus) [inline changelog/12231.feature.rst:13]: Nice!

IMHO this explains the history a bit better (mostly reordered the items):

```suggestion
With pytest 8.0, `-rx` or `-ra` would not only turn on summary reports for xfail, but also report the tracebacks for xfail results. This caused issues with some projects that utilize xfail, but don't want to see all of the xfail tracebacks.

This change detaches xfail tracebacks from `-rx`, and now we turn on xfail tracebacks with `--xfail-tb`. With this, the default `-rx`/ `-ra` behavior is identical to pre-8.0 with respect to xfail tracebacks. While this is a behavior change, it brings default behavior back to pre-8.0.0 behavior, which ultimately was considered the better course of action.
```

File: src/_pytest/terminal.py [hunk 3]
```diff
@@ -1086,21 +1093,28 @@ def print_teardown_sections(self, rep: TestReport) -> None:
...
+        style = self.config.option.tbstyle if show_tb else "no"
+        self.summary_failures_combined("xfailed", "XFAILURES", style=style)
 
     def summary_failures_combined(
-        self, which_reports: str, sep_title: str, needed_opt: Optional[str] = None
+        self,
+        which_reports: str,
+        sep_title: str,
+        needed_opt: Optional[str] = None,
+        style: Optional[str] = None,
```
Comment (nicoddemus) [inline src/_pytest/terminal.py:1109]: I think it makes sense to always require `style`, as it is only called from two places and those always pass `style`.

```suggestion
        sep_title: str,
        *,
        style: str,
        needed_opt: Optional[str] = None,        
```

File: testing/test_terminal.py [hunk 4]
```diff
@@ -2909,54 +2909,75 @@ def test_xfail_reason():
...
     result.stdout.fnmatch_lines(
         [
+            "*= FAILURES =*",
+            "*test_xfail_tb_line.py:5: assert 1 == 2",
             "*= XFAILURES =*",
-            "*test_xfail_tb_line.py:6: assert 1 == 2",
-            "*= short test summary info =*",
-            "XFAIL test_xfail_tb_line.py::test_xfail",
-            "*= 1 xfailed in * =*",
+            "*test_xfail_tb_line.py:10: assert 3 == 4",
```
Comment (nicoddemus) [inline testing/test_terminal.py:2980]: Let's also match `short test summary info` to ensure the summary happened before it.

```suggestion
            "*test_xfail_tb_line.py:10: assert 3 == 4",
            "*short test summary info*",
```

File: testing/test_terminal.py [hunk 5]
```diff
@@ -2909,54 +2909,75 @@ def test_xfail_reason():
...
-            "> *assert a == b*",
-            "E *assert 1 == 2*",
-            "test_summary_xfail_tb.py:6: AssertionError*",
-            "*= short test summary info =*",
-            "XFAIL test_summary_xfail_tb.py::test_xfail",
-            "*= 1 xfailed in * =*",
+            "*def test_xfail():*",
+            "*        c, d = 3, 4*",
+            "*>       assert c == d*",
+            "*E       assert 3 == 4*",
```
Comment (nicoddemus) [inline testing/test_terminal.py:2966]:
```suggestion
            "*E       assert 3 == 4*",
            "*short test summary info*",
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Looks great, thanks @okken!

I have left some minor requests, please take a look.

**Comment by okken** [comment]
> Looks great, thanks @okken!
> 
> I have left some minor requests, please take a look.

Thanks @nicoddemus. I like all the suggestions.

**Comment by okken** [comment]
Let me know if there's anything else needed.

**Review by nicoddemus — APPROVED** [review]
Thanks, looks great! 

I will let it open for a few more days to give others a chance of reviewing it too. 👍

**Comment by nicoddemus** [comment]
(Note: we should squash/merge this)

**Review by bluetech — APPROVED** [review]
LGTM, thanks!

**Comment by okken** [comment]
@nicoddemus Squash and merge ok now? Or wanna wait a bit more?

**Comment by nicoddemus** [comment]
Thanks @okken!

**Pull request merged**